### PR TITLE
Handles symlinks pointing to directories

### LINKED ISSUE #174: Crashes when directory contains symlink to a directory. (When `append_dir_all`.)
Hi @alexcrichton! Respect your hard works on Rust!

I've found a bug for the tar crate. This is a small project to re-create the error.

https://github.com/tbrand/tar-error

The error occurs when source directory contains a symlink **to a directory**.
It might be related [this issue](https://github.com/alexcrichton/tar-rs/issues/147).
I hope this will be fixed soon. :smile:

Thanks!

### PR DESCRIPTION
**Pull request by boastful-squirrel**
This is an attempt to fix #174 and #147 

Note that I'm quite a beginner so I'm open to any criticism on the code quality or anything else. I tested my code on a directory containing a symlink to a directory, containing a symlink to a directory. I'm pretty confident my code can be optimised so it matches the first condition, though I can't seem to a find a predicate that would match... (`if is_dir || ????`).

If you need me to write unit tests associated with this features, I can do it

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! I was thinking the same thing though that ideally the refactored `browse_dir` function would only be called from one location because it seems odd for a non-file to leak into the "this is a file" clause later on.

Additionally yeah tests for this would be greatly appreciated!

**Comment by boastful-squirrel** [comment]
So I think this should do it... It passes the tests, and my use case. I'll try to write tests, I never did so I can't promise anything.

**Comment by boastful-squirrel** [comment]
So I wrote a very basic test, as I said I never did it before so I'm open to any comments. This test fails on `master`, and succeeds on this branch.

File: src/builder.rs [hunk 1]
```diff
@@ -489,7 +489,7 @@ fn append_dir_all(
     let mut stack = vec![(src_path.to_path_buf(), true, false)];
     while let Some((src, is_dir, is_symlink)) = stack.pop() {
         let dest = path.join(src.strip_prefix(&src_path).unwrap());
-        if is_dir {
+        if is_dir || is_symlink && follow && src.is_dir() {
```
Comment (alexcrichton) [inline src/builder.rs]: Could parentheses be used here around the added clause to explicitly show the precedence? Additionally could you add a comment as well saying that `is_dir` being false is about the symlink, and the call to `.is_dir()` reaches through the symlink?

**Comment by alexcrichton** [comment]
Looks great! Just a few minor nits but otherwise looks good to go to me

**Comment by boastful-squirrel** [comment]
Hey, here you go :)

**Comment by alexcrichton** [comment]
Thanks!

**Comment by boastful-squirrel** [comment]
Do you plan on releasing 0.4.22 with this fix by the way ?

**Comment by alexcrichton** [comment]
Done!

**Pull request merged**